src/overlay/SurveyManager.cpp

```diff
@@ -20,6 +20,23 @@ namespace stellar
 
 uint32_t const SurveyManager::SURVEY_THROTTLE_TIMEOUT_MULT(3);
 
+// Function to call when something has gone wrong. Throws an exception with
+// exceptionMessage in testing, otherwise logs an error (errorMessage) and a
+// note to report the bug.
+template <typename... Ts>
+static void
+emitInconsistencyError(const char* exceptionMessage,
+                       std::string_view errorMessage, Ts... errorExtra)
+{
+    static_assert(std::conjunction_v<std::is_same<std::string_view, Ts>...>);
+#ifdef BUILD_TESTS
+    throw std::runtime_error(exceptionMessage);
+#else
+    CLOG_ERROR(Overlay, errorMessage, errorExtra...);
+    CLOG_ERROR(Overlay, "{}", REPORT_INTERNAL_BUG);
+#endif
+}
```

Comment:
Thinking about it a little more, this is a potentially useful function elsewhere, and it's also very close to `emitInconsistencyError` in `SurveyMessageLimiter`. Can you refactor this out to somewhere more accessible (maybe `util/GlobalChecks.{h,cpp}`), perhaps also with a more descriptive name? We won't be able to get rid of `SurveyMessageLimiter::emitInconsistencyError` because it needs to call `reset()`, but it could call this new function followed by `reset`.

Reply:
One additional consideration is that the only other place in the codebase where we currently do an `#ifdef BUILD_TESTS` immediately followed by a `throw` is [src/overlay/Flowcontrol.cpp:256](https://github.com/drebelsky/stellar-core/blob/b23dd03b1458492bc43646cb2c8aef1dfede977e/src/overlay/FlowControl.cpp#L256). The logging module having the first parameter be a macro token that gets concatenated makes it a little more annoying to generalize outside of Overlay (could, e.g., be done with a macro). That is, I think the cleanest/smallest interface change would be to make this an `overlay` util, but I'm happy to make it a global util if you thinks that's better.

Reply:
Keeping it within overlay sounds like a fine plan to me!